Added types definition for "level" npm package

danielfx90 opened the pull request:
## Template
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

danielfx90 commented:
Hi @MeirionHughes @danwbyrne, I saw you were the ones that defined the types for packages:
* [abstract-leveldown](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/abstract-leveldown)
* [encoding-down](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/encoding-down)
* [level-codec](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/level-codec)
* [levelup](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/levelup) (+@carsonfarmer)

Would you mind having a look at this PR?

Thanks!

peterblazejewicz commented on types/level/tslint.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "dtslint/dt.json",
+  "rules": {
+    "no-outside-dependencies": false
```
please move this inline `// tslint:disable .....`, so you could put a comment why this is needed

peterblazejewicz commented on types/level/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+
+declare function level(location: string, options?: AbstractOptions, callback?: ErrorCallback): LevelDB;
+
+export default level;
```
IMO this look like proper CommonJS package (4 years old export):
https://github.com/Level/level/blob/master/level.js#L1.
This is not ESM (though it could be used as such when consuming types)
```suggestion
export = level;
```

danielfx90 replied:
@peterblazejewicz I committed your suggestions but now I realize that I also have the `LevelDB` interface exported. With `export = level;` I cannot have anything else exported.

How do you recommend solving this issue?

danielfx90 replied:
Nevermind, I found out how. Sorry

peterblazejewicz commented on types/level/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+import { LevelUp } from 'levelup';
+
+export interface LevelDB<K = any, V = any> extends LevelUp<EncodingDown<K, V>> {
+  errors: /* typeof levelerrors */ any; // ? level-errors is not in DT
```
this could be just:
```ts
errors: Record<string, Error>;
```
(Liskov substitution, reversed), the list of errors can be an union (`NotFoundError` | `....`). This should be done upstream (levelup) though

peterblazejewicz commented on types/level/level-tests.ts:
```diff
@@ -0,0 +1,23 @@
+import level, { LevelDB } from 'level';
```
```suggestion
import level = require('level');
import { LevelDB } from 'level';
```

peterblazejewicz commented:
@danielfx90 fancy running prettier on this? To use DT defaults (like quoting):
https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

danielfx90 commented:
@peterblazejewicz oh i'm terribly sorry. I must have used my personal vscode config by mistake.
I'll definitely fix it.

amcasey reviewed (changes requested):
Please address @peterblazejewicz's feedback

danielfx90 commented:
@peterblazejewicz @amcasey all done!

Changelog since last review:

- Enabled again `no-outside-dependencies` rule. Turns out it was only failing on my computer. CI runs it fine.
- Applied suggestions for import/export statements aligning with CommonJS.
- Added proper typing for `errors` property in `levelup` and re-used the definition in `level`.
- Applied prettier formatting to `level` and `levelup` (since I changed `errors` property upstream).

Thank you for the feedback and collaboration!

peterblazejewicz commented:
@danielfx90 I meant changes only in the new package :) I'd probably avoid changing format in the LevelUp
For me it's OK, but now we need opinion of those folks, thx!
@MeirionHughes @danwbyrne @carsonfarmer

danielfx90 commented:
@peterblazejewicz OK! Reverted prettier formatting over Levelup files. Kept only the `errors` definition so that it's easier for @MeirionHughes @danwbyrne @carsonfarmer

Thanks!

amcasey commented:
LGTM but let's give the owners some time.

amcasey reviewed (commented):
Not signing off to prevent the bot from jumping the gun.

carsonfarmer commented:
FWIW I'm cool with the updates! 👍 from me.

amcasey reviewed (approved):
Thanks, all!

danielfx90 commented:
Ready to merge

danielfx90 commented:
Wooo-hoooo! Thanks @peterblazejewicz @amcasey @carsonfarmer for the guidance and feedback!
